core/data-plane/data-plane-util/src/main/java/org/eclipse/edc/connector/dataplane/util/sink/ParallelSink.java

```diff
@@ -64,7 +64,7 @@ public CompletableFuture<StatusResult<Void>> transfer(DataSource source) {
                     .exceptionally(throwable -> StatusResult.failure(ERROR_RETRY, "Unhandled exception raised when transferring data: " + throwable.getMessage()));
         } catch (Exception e) {
             monitor.severe("Error processing data transfer request: " + requestId, e);
-            return CompletableFuture.completedFuture(StatusResult.failure(ERROR_RETRY, "Error processing data transfer request"));
+            return CompletableFuture.completedFuture(StatusResult.failure(ERROR_RETRY, String.format("Error processing data transfer request - %s", e.getMessage())));
```

Comment:
This way the exception message will be propagated to the counter-part, I'm not sure this is something we really want.

Reply:
Agreed, we don't want this. It's sufficient to send the exception to the monitor. We could discuss a correlation mechanism where a GUID is generated on an exception, sent to the monitor with the exception, and the GUID returned to the client.

Reply:
I can understand the reason why we don't want to propagate the message retrieved by the DataSource but it would always cause the consumer counter-part end user to first believe it is their connector failure. I think a GUID on exceptions is a great approach we could discuss.

Reply:
The client might/should not be aware of external dependencies and a status 500 combined with an undetailed message would certainly cause this misinterpretation, such as the one in the report that originated the issue.

I think we could improve the exception handling to retrieve correlation information as you suggested.

Reply:
The client *should* interpret the problem as an issue with the system it communicates with (the data plane), even if another system behind it caused the error. 

The issue is not exception handling; that is done correctly in the code. There are two issues as I see it. 

The first is providing a correlation mechanism for a *data plane operator* (not the client) to link unsuccessful responses returned to a client and internal diagnostic information. IMO, this can be handled by a Decision Record that calls for two minor changes. Specifically:

1. One or more correlation ids are sent to the client
2. Exceptional conditions sent to the monitor should contain a correlation id

A second issue is the HTTP DataSource could log an error response body (if one exists).

Reply:
> ...
> 
> 1. One or more correlation ids are sent to the client
> 2. Exceptional conditions sent to the monitor should contain a correlation id
> 
> A second issue is the HTTP DataSource could log an error response body (if one exists).

- `ParallelSink`'s implementation already provides something very similar to what you suggested. It uses the DataFlowRequest id to correlate the request to the exception thrown, it logs the id alongside the exception message and returns just the id to the client, and it also enables the operator to link unsuccessful responses to internal diagnostic information. Wouldn't it be the case just to reproduce these actions also in `OutputStreamDataSink`?

Reply:
> > ...
> > 
> > 1. One or more correlation ids are sent to the client
> > 2. Exceptional conditions sent to the monitor should contain a correlation id
> > 
> > A second issue is the HTTP DataSource could log an error response body (if one exists).
> 
> * `ParallelSink`'s implementation already provides something very similar to what you suggested. It uses the `DataFlowRequest` id to correlate the request to the exception thrown, it logs the id alongside the exception message and returns just the id to the client, and it also enables the operator to link unsuccessful responses to internal diagnostic information. Wouldn't it be the case just to reproduce these actions also in `OutputStreamDataSink`?

Yes, I think using the request id would be fine.